### PR TITLE
:sparkles: Refactor MS controller status patching

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:

This PR improves the MachineSet code by:

- Moves all the code to sync Status to machineset_controller.go.
- Uses patch instead of update for the status resource, but fall back to Update if the validation fails (explained in comment).
- Moves setting Status.Selector in `calculateStatus` instead of the main reconciler function.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #1689

### CODE REVIEW

File: controllers/machineset_controller.go [hunk 1]
```diff
@@ -549,3 +542,118 @@ func (r *MachineSetReconciler) hasMatchingLabels(machineSet *clusterv1.MachineSe
...
+	// template of the replica set, the matching machines may have more
+	// labels than are in the template. Because the label of machineTemplateSpec is
+	// a superset of the selector of the replica set, so the possible
+	// matching machines must be part of the filteredMachines.
+	fullyLabeledReplicasCount := 0
+	readyReplicasCount := 0
+	availableReplicasCount := 0
+	templateLabel := labels.Set(ms.Spec.Template.Labels).AsSelectorPreValidated()
+
+	// Retrieve Cluster, if any.
```
Comment (ncdc) [inline controllers/machineset_controller.go]: Remove the "if any" part? Or handle the error?

Comment (vincepri) [reply controllers/machineset_controller.go]: This was copied from the other file, I have no idea why it ignores the error though

Comment (detiber) [reply controllers/machineset_controller.go]: We used to have a fallback to using the in-cluster client when the Cluster wasn't defined: https://github.com/kubernetes-sigs/cluster-api/blob/release-0.2/controllers/machineset_status.go#L141-L146 

I suspect this call was never updated when that was removed.

Comment (vincepri) [reply controllers/machineset_controller.go]: I’ll remove it in a bit

Comment (vincepri) [reply controllers/machineset_controller.go]: Done, removed the line, and passed in the cluster from the caller

File: controllers/machineset_controller.go [hunk 2]
```diff
@@ -589,3 +582,115 @@ func (r *MachineSetReconciler) hasMatchingLabels(machineSet *clusterv1.MachineSe
...
 	return !util.HasOwner(ms.OwnerReferences, clusterv1.GroupVersion.String(), []string{"MachineDeployment", "Cluster"})
 }
+
+func (r *MachineSetReconciler) calculateStatus(cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, filteredMachines []*clusterv1.Machine) clusterv1.MachineSetStatus {
+	logger := r.Log.WithValues("machineset", ms.Name, "namespace", ms.Namespace)
+	newStatus := ms.Status
+
+	// Copy label selector to its status counterpart in string format.
+	// This is necessary for CRDs including scale subresources.
+	selector, _ := metav1.LabelSelectorAsSelector(&ms.Spec.Selector)
```
Comment (detiber) [inline controllers/machineset_controller.go]: Is it actually safe to ignore the error here? I know we have a validating webhook now, but that wouldn't affect pre-existing stored resources (unless they have been read and then re-written to the new storage version).

Comment (vincepri) [reply controllers/machineset_controller.go]: I figured given that we're coming from a `LabelSelector` type that this function should never return an error, unless something very wrong happened.

I'll add it though, just to be safe

**Pull request merged**